xarray/core/dataset.py

```diff
@@ -8678,16 +8678,23 @@ def _integrate_one(self, coord, datetime_unit=None, cumulative=False):
                     coord_names.add(k)
             else:
                 if k in self.data_vars and dim in v.dims:
+                    # cast coord data to duck array if needed
+                    if isinstance(v.data, array_type("cupy")):
+                        coord_data = duck_array_ops.get_array_namespace(v.data).asarray(
+                            coord_var.data
+                        )
+                    else:
+                        coord_data = coord_var.data
```

Comment:
It looks like this could be encapsulated into an `as_like_array()` helper function?

I'd like to keep the explicit array namespace stuff out of most xarray functions.